anndata/_io/specs/registry.py

```diff
@@ -17,6 +21,108 @@ class IOSpec(NamedTuple):
+        self.key = key
+        self.registry = self._hierarchical_registry()
+        super().__init__(str(self))
+
+    def _hierarchical_registry(
+        self,
+    ) -> Mapping[type, Mapping[IOSpec | type | tuple[type, str], Set[frozenset[str]]]]:
+        reg = {}
+        for typ, spec_or_src_type, modifiers in getattr(_REGISTRY, self.registry_name):
+            reg.setdefault(typ, {}).setdefault(spec_or_src_type, set()).add(modifiers)
+        return reg
+
+    def __str__(self) -> str:
+        return f"No such {self.registry_name} function registered: {self._get_msg()}."
+
+    @property
+    def typ(self) -> type:
+        return self.key[0]
+
+    @property
+    def spec_or_dest_type(self) -> IOSpec | type | tuple[type, str]:
+        return self.key[1]
+
+    @property
+    def modifiers(self) -> frozenset[str]:
+        return self.key[2]
+
+    def _get_msg(self) -> str:
+        if self.typ not in self.registry:
+            return f"Unknown {self.type_desc}"
+
+        if self.spec_or_dest_type not in self.registry[self.typ]:
+            return self._get_spec_or_dest_type_msg()
+
+        if self.modifiers not in self.registry[self.typ][self.spec_or_dest_type]:
+            desc = self._get_modifiers_desc()
+            return f"Unknown modifier set {self.modifiers} for {desc}"
+
+        assert False, f"Don’t create NoSuchIO error from valid key: {self.key}"
+
+    @property
+    @abstractmethod
+    def type_desc(self) -> str:
+        ...
+
+    @abstractmethod
+    def _get_spec_or_dest_type_msg(self) -> str:
+        ...
+
+    @abstractmethod
+    def _get_modifiers_desc(self) -> str:
+        ...
+
+
+class NoSuchWrite(NoSuchIO):
+    @property
+    def dest_type(self) -> type | tuple[type, str]:
+        return self.spec_or_dest_type
+
+    @property
+    def type_desc(self) -> str:
+        return f"source type {self.typ.__name__}"
+
+    def _get_spec_or_dest_type_msg(self) -> str:
+        return f"Destination type {self.dest_type} not found for {self.type_desc}"
+
+    def _get_modifiers_desc(self) -> str:
+        return f"{self.type_desc} and destination type {self.dest_type}"
+
+
+class NoSuchRead(NoSuchIO):
+    @property
+    def spec(self) -> IOSpec:
+        return self.spec_or_dest_type
+
+    @property
+    def type_desc(self) -> str:
+        return f"destination type {self.typ.__name__}"
+
+    def _get_spec_or_dest_type_msg(self) -> str:
+        enc_types = {spec.encoding_type for spec in self.registry[self.typ]}
+        if self.spec.encoding_type not in enc_types:
+            return (
+                f"Unknown encoding type “{self.spec.encoding_type}” "
+                f"for {self.type_desc}"
+            )
+        return (
+            f"Unknown encoding version {self.spec.encoding_version} "
+            f"for {self.type_desc}’s encoding “{self.spec.encoding_type}”"
+        )
+
+    def _get_modifiers_desc(self) -> str:
+        return f"{self.type_desc}’s encoding {self.spec}"
+
+
```

Comment:
I agree this is too verbose. I think you could do something more like this:

```suggestion
class NoSuchIO(KeyError):
    pass


class NoSuchWrite(NoSuchIO):

    @staticmethod
    def _from_parts(cls, dest_type, typ, modifiers) -> NoSuchWrite:
        msg = f"No method registered for writing {typ} into {dest_type}"
        if modifiers:
            msg += f" with {modifiers}"
        return cls(msg)


class NoSuchRead(NoSuchIO):

    @staticmethod
    def _from_parts(cls, method: str, registry, src_typ, spec) -> NoSuchRead:
        versions = [s.encoding_version for s in self.registry[method] if spec.encoding_type == s.encoding_type]
        msg = f"No {method} method registered for {spec} from {src_typ}"
        if versions:
            msg += f", readers only found for versions: {versions}"
        return cls(msg)
```

BTW, the way of doing look ups for reading methods needs to change (for cases where forward compatible reading is possible) so I wouldn't hook into that logic much.

Reply:
Also IDK if we need three classes here.

Reply:
The tests are now complete. Feel free to shorten my code if you think it’s worth the less expressive error messages.